### PR TITLE
Dynamic queries type plumbing

### PR DESCRIPTION
**Pull request by Cypher1**

### CODE REVIEW

File: src/runtime/schema.ts [hunk 1]
```diff
@@ -174,6 +173,9 @@ export class Schema {
         return false;
       }
     }
+    if (Refinement.isAtleastAsSpecificAs(this.refinement, otherSchema.refinement) === AtleastAsSpecific.NO) {
```
Comment (Cypher1) [inline src/runtime/schema.ts]: It seems we were only checking the field level handles previously. @ragavsachdeva can you double check for me that this isn't duplicate work?

File: src/runtime/schema.ts [hunk 2]
```diff
@@ -217,13 +219,16 @@ export class Schema {
   toInlineSchemaString(options?: {hideFields?: boolean}): string {
     const names = this.names.join(' ') || '*';
     const fields = Object.entries(this.fields).map(Schema.fieldToString).join(', ');
-    return `${names} {${fields.length > 0 && options && options.hideFields ? '...' : fields}}`;
+    return `${names} {${fields.length > 0 && options && options.hideFields ? '...' : fields}}${this.refinement ? this.refinement.toString() : ''}`;
```
Comment (Cypher1) [inline src/runtime/schema.ts]: I've moved the to String for the refinement to the schema level to ensure that it is always available (and reduce the particle handle connection's complexity.

File: src/runtime/schema.ts (hunk 1 above)
Comment (ragavsachdeva) [reply src/runtime/schema.ts]: With the current implementation this is not needed. The system cannot handle multivariate case and will return AtleastAsSpecific.UNKNOWN. In the case where the expression is univariate, it is propagated to field level and is handled there.

However, I added this exact same code in the PR I'm working on where we disable field level refinements. I believe it is better to add this functionality as part of that.

File: src/runtime/refiner.ts [hunk 3]
```diff
@@ -158,7 +142,8 @@ export class Refinement {
       // Therefore, we TRY to rearrange, if possible.
       this.expression = this.expression.rearrange();
     } catch (e) {
-      console.log(e);
+      // tslint:disable-next-line no-empty
```
Comment (ragavsachdeva) [inline src/runtime/refiner.ts]: Don't need this try/catch anymore

File: src/runtime/refiner.ts [hunk 4]
```diff
@@ -526,6 +551,10 @@ export class FieldNamePrimitive extends RefinementExpression {
     return new Set<string>([this.value]);
   }
 
+  getQueryNames(): Set<string> {
+    return new Set<string>();
+  }
```
Comment (shans) [inline src/runtime/refiner.ts]: provide default impl on superclass rather than repeating this in each subclass.

Comment (Cypher1) [reply src/runtime/refiner.ts]: Done.

P.S. We have a bunch of visitor patterns in our code base, it would be nice to work out a higher order interface that we could implement for them.

Then we could add `T implements Visitable<T>` to the type and something like
`children(): T[] { return [self.rightExpr self.rightExpr]}` in the nodes
Giving us

```
getQueryNames(): Set<string> {
    return this.visit(x => x.getQueryNames(), merge=Visitor.unionSets);
}

eval(): Set<string> {
    return this.visit(x => x.eval(), merge=this.applyOperator)
}
```

TLDR: Josh would like more Higher order types to avoid code duplication and implementation burden.
